### PR TITLE
CPM-997: Handle several identifier generators for subscribers

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

File: components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php [hunk 1]
```diff
@@ -69,17 +69,28 @@ public function setIdentifier(GenericEvent $event): void
...
+                    $product->getFamily()?->getCode(),
+                    $this->flatValues($product),
+                );
+                if ($identifierGenerator->match($productProjection)) {
+                    $targetMatched = true;
+
+                    try {
+                        $this->setGeneratedIdentifier($identifierGenerator, $productProjection, $product);
+                    } catch (UnableToSetIdentifierException $e) {
+                        // TODO Check ce que ça fait quand y'a 2 generateurs invalides ?
```
Comment (anaelChardan) [inline components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php:91]: Et ca fait quoi alors?

Comment (pierallard) [reply components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php:91]: MAIS NANOU MAIS ENFIN NANOU

Comment (anaelChardan) [reply components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php]: OUI PIERRRRRRRE

File: components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php [hunk 2]
```diff
@@ -69,17 +69,27 @@ public function setIdentifier(GenericEvent $event): void
...
+                    $this->flatValues($product),
+                );
+                if ($identifierGenerator->match($productProjection)) {
+                    $targetMatched = true;
+
+                    try {
+                        $this->setGeneratedIdentifier($identifierGenerator, $productProjection, $product);
+                    } catch (UnableToSetIdentifierException $e) {
+                        $this->eventDispatcher->dispatch(new UnableToSetIdentifierEvent($e));
+                    }
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php:92]:
```suggestion
                    } finally {
                        // we don't want to try to generate any new value for this target identifier
                        break;
                    }
```

File: components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php [hunk 3]
```diff
@@ -69,17 +69,27 @@ public function setIdentifier(GenericEvent $event): void
...
-                $this->flatValues($product),
-            );
-            if ($identifierGenerator->match($productProjection)) {
-                try {
-                    $this->setGeneratedIdentifier($identifierGenerator, $productProjection, $product);
-                } catch (UnableToSetIdentifierException $e) {
-                    $this->eventDispatcher->dispatch(new UnableToSetIdentifierEvent($e));
+        $productProjection = null;
+        foreach ($this->getIdentifierGeneratorsByTarget() as $identifierGeneratorsByTarget) {
+            $targetMatched = false;
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php:74]:
```suggestion
```

File: components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php [hunk 4]
```diff
@@ -69,17 +69,27 @@ public function setIdentifier(GenericEvent $event): void
...
+            foreach ($identifierGeneratorsByTarget as $identifierGenerator) {
+                if ($targetMatched) {
+                    continue;
+                }
+
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Subscriber/SetIdentifiersSubscriber.php:79]:
```suggestion
```

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php [hunk 5]
```diff
@@ -155,8 +155,12 @@ public function createAction(Request $request)
 
             $events = $this->unableToSetIdentifiersSubscriber->getEvents();
             if (\count($events) > 0) {
-                $normalizedProduct['meta']['identifier_generator_warnings'] =
-                    $this->unableToSetIdentifierExceptionPresenter->present($events[0]->getException());
+                $normalizedProduct['meta']['identifier_generator_warnings'] = \array_merge(
+                    ...\array_map(
+                        fn ($event): array => $this->unableToSetIdentifierExceptionPresenter->present($event->getException()),
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php:160]:
```suggestion
                        fn (UnableToSetIdentifierEvent $event): array => $this->unableToSetIdentifierExceptionPresenter->present($event->getException()),
```

**Review by mmetayer — APPROVED** [review]
Only suggestions from personal preference, you can choose whether you want to apply the or not

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php (hunk 5 above)
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php:160]: I did this on purpose to not add coupling.
If you want me to do that, we need to expose this Event in our API. Is it what we want?

**Pull request merged**